File: supabase/functions/_shared/lecithin.ts
// ================================================================
// PureMark Edge Functions - Lecithin Source Detection
// Language-aware lecithin detection for halal and allergen analysis
// ================================================================

import { normalizeText, containsEnumber } from "./helpers.ts";

// Lecithin source mapping - keys are terms that indicate a specific source
export const LECITHIN_SOURCES: Record<string, string> = {
  // Sunflower variants (multiple languages)
  "sunflower": "sunflower",
  "tournesol": "sunflower",      // French
  "sonnenblume": "sunflower",    // German
  "girasol": "sunflower",        // Spanish
  "girasole": "sunflower",       // Italian

  // Soy variants (multiple languages)
  "soy": "soy",
  "soya": "soy",
  "soja": "soy",                 // French/German/Spanish

  // Rapeseed / Canola variants
  "rapeseed": "rapeseed",
  "colza": "rapeseed",           // French
  "raps": "rapeseed",            // German
  "canola": "rapeseed",

  // Egg variants
  "egg": "egg",
  "oeuf": "egg",                 // French
  "ei": "egg",                   // German
  "huevo": "egg",                // Spanish
  "uovo": "egg",                 // Italian
};

// Lecithin halal status by source
export const LECITHIN_HALAL_STATUS: Record<string, string> = {
  "sunflower": "HALAL",           // Plant-based, inherently halal
  "soy": "UNVERIFIED",            // Plant-based BUT processing may involve alcohol
  "rapeseed": "HALAL",            // Plant-based, inherently halal
  "egg": "HALAL",                 // Halal (eggs are halal), but allergen
  "unspecified": "UNVERIFIED",    // Unknown source, needs verification
};

// Lecithin allergen mapping
export const LECITHIN_ALLERGENS: Record<string, string> = {
  "soy": "soy",
  "egg": "egg",
};

export interface LecithinDetectionResult {
  isLecithin: boolean;
  source: string | null;
  explanation: string | null;
}

/**
 * Detect if ingredient contains lecithin and identify its source.
 *
 * @param ingredientText - The ingredient text to analyze
 * @returns Detection result with source and explanation
 */
export function detectLecithinSource(ingredientText: string): LecithinDetectionResult {
  const text = normalizeText(ingredientText);

  // Check if this ingredient contains lecithin
  const lecithinTerms = [
    "lecithin", "lecithine", "lécithine", "lecitina",
    "e322"
  ];

  const isLecithin = lecithinTerms.some(term => text.includes(term)) || containsEnumber(text, "322");

  if (!isLecithin) {
    return { isLecithin: false, source: null, explanation: null };
  }

  // Try to identify the source
  let detectedSource: string | null = null;

  for (const [sourceTerm, sourceType] of Object.entries(LECITHIN_SOURCES)) {
    if (text.includes(sourceTerm)) {
      detectedSource = sourceType;
      break;
    }
  }

  // If no specific source found, it's unspecified
  if (detectedSource === null) {
    return {
      isLecithin: true,
      source: "unspecified",
      explanation: "Lecithin with unspecified source"
    };
  }

  // Build explanation
  const sourceNames: Record<string, string> = {
    "sunflower": "Sunflower lecithin",
    "soy": "Soy lecithin",
    "rapeseed": "Rapeseed/canola lecithin",
    "egg": "Egg lecithin",
  };

  const explanation = `${sourceNames[detectedSource] || detectedSource.charAt(0).toUpperCase() + detectedSource.slice(1) + ' lecithin'} detected`;

  return { isLecithin: true, source: detectedSource, explanation };
}

/**
 * Return allergens associated with a lecithin source.
 * Only soy lecithin and egg lecithin are allergens.
 */
export function getLecithinAllergens(lecithinSource: string): string[] {
  if (lecithinSource in LECITHIN_ALLERGENS) {
    return [LECITHIN_ALLERGENS[lecithinSource]];
  }
  return [];
}

/**
 * Detect lecithin source by checking BOTH original and normalized text.
 * Fixes issues where GPT incorrectly normalizes source terms.
 */
export function detectLecithinSourceCombined(
  originalText: string,
  normalizedText: string
): LecithinDetectionResult {
  // First check the original text (more reliable for source detection)
  const origResult = detectLecithinSource(originalText);

  // Then check normalized text
  const normResult = detectLecithinSource(normalizedText);

  // If neither contains lecithin, return early
  if (!origResult.isLecithin && !normResult.isLecithin) {
    return { isLecithin: false, source: null, explanation: null };
  }

  // Prioritize original text source if it's specific (not unspecified)
  if (origResult.isLecithin && origResult.source && origResult.source !== "unspecified") {
    return {
      isLecithin: true,
      source: origResult.source,
      explanation: `${origResult.explanation} (from original text)`
    };
  }

  // Fall back to normalized text source
  if (normResult.isLecithin && normResult.source && normResult.source !== "unspecified") {
    return {
      isLecithin: true,
      source: normResult.source,
      explanation: `${normResult.explanation} (from normalized text)`
    };
  }

  // If both are lecithin but source is unspecified in both
  if (origResult.isLecithin || normResult.isLecithin) {
    return {
      isLecithin: true,
      source: "unspecified",
      explanation: "Lecithin with unspecified source"
    };
  }

  return { isLecithin: false, source: null, explanation: null };
}

File: backend/services/lecithin.py
# ================================================================
# PureMark Backend - Lecithin Detection Service
# Detect lecithin source for accurate halal/allergen classification
# ================================================================

import re
from typing import Dict, Any

# Lecithin source patterns
LECITHIN_PATTERNS = {
    "sunflower": [
        r"sunflower\s*lecithin",
        r"lecithin\s*\(?\s*sunflower\s*\)?",
        r"lecithine?\s*(?:de\s*)?(?:tournesol|girasol)",  # French/Spanish
        r"sonnenblumen\s*lecithin",  # German
    ],
    "soy": [
        r"soy\s*lecithin",
        r"soya\s*lecithin",
        r"lecithin\s*\(?\s*soy(?:a)?\s*\)?",
        r"lecithine?\s*(?:de\s*)?soja",  # French/Spanish
        r"soja\s*lecithin",  # German
    ],
    "rapeseed": [
        r"rapeseed\s*lecithin",
        r"canola\s*lecithin",
        r"lecithin\s*\(?\s*(?:rapeseed|canola)\s*\)?",
        r"lecithine?\s*(?:de\s*)?colza",  # French
        r"raps\s*lecithin",  # German
    ],
    "egg": [
        r"egg\s*lecithin",
        r"lecithin\s*\(?\s*egg\s*\)?",
        r"lecithine?\s*(?:d['']?\s*)?(?:oeuf|huevo)",  # French/Spanish
    ],
}

# Generic lecithin patterns (source unknown)
GENERIC_LECITHIN = [
    r"\blecithin\b",
    r"\blecithine?\b",
    r"\be322\b",
    r"\be\s*322\b",
]

# Halal status by lecithin source
LECITHIN_HALAL_STATUS = {
    "sunflower": "HALAL",
    "soy": "UNVERIFIED",  # May involve alcohol processing
    "rapeseed": "HALAL",
    "egg": "HALAL",
    "unspecified": "UNVERIFIED",
}


def detect_lecithin_source(text: str) -> Dict[str, Any]:
    """
    Detect if text contains lecithin and identify its source.

    Args:
        text: Text to search for lecithin

    Returns:
        Dict with is_lecithin, source, halal_status
    """
    if not text:
        return {"is_lecithin": False, "source": None, "halal_status": None}

    text_lower = text.lower()

    # Check for specific sources first
    for source, patterns in LECITHIN_PATTERNS.items():
        for pattern in patterns:
            if re.search(pattern, text_lower):
                return {
                    "is_lecithin": True,
                    "source": source,
                    "halal_status": LECITHIN_HALAL_STATUS.get(source, "UNVERIFIED")
                }

    # Check for generic lecithin
    for pattern in GENERIC_LECITHIN:
        if re.search(pattern, text_lower):
            return {
                "is_lecithin": True,
                "source": "unspecified",
                "halal_status": "UNVERIFIED"
            }

    return {"is_lecithin": False, "source": None, "halal_status": None}


def detect_lecithin_source_combined(original_text: str, normalized_text: str) -> Dict[str, Any]:
    """
    Detect lecithin source from multiple text sources.

    Args:
        original_text: Original text (may be in foreign language)
        normalized_text: Normalized/translated text

    Returns:
        Dict with is_lecithin, source, halal_status
    """
    # Try original text first
    result = detect_lecithin_source(original_text)
    if result["is_lecithin"] and result["source"] != "unspecified":
        return result

    # Try normalized text
    result_norm = detect_lecithin_source(normalized_text)
    if result_norm["is_lecithin"] and result_norm["source"] != "unspecified":
        return result_norm

    # Return whatever we found
    if result["is_lecithin"]:
        return result
    return result_norm
